hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java

```diff
@@ -397,75 +408,66 @@ private void startInstant() {
-   *
-   * <p>Starts a new instant, a writer can not flush data buffer
-   * until it finds a new inflight instant on the timeline.
    */
-  private void initInstant(String instant) {
+  private void recommitInstant(long checkpointId, String instant, WriteMetadataEvent[] bootstrapBuffer) {
     HoodieTimeline completedTimeline = this.metaClient.getActiveTimeline().filterCompletedInstants();
     if (instant.equals(WriteMetadataEvent.BOOTSTRAP_INSTANT) || completedTimeline.containsInstant(instant)) {
       // the last instant committed successfully
-      reset();
+      // no-op
     } else {
       LOG.info("Recommit instant {}", instant);
-      // Recommit should start heartbeat for lazy failed writes clean policy to avoid aborting for heartbeat expired.
+      // Recommit should start heartbeat for lazy failed writes clean policy to avoid aborting for heartbeat expired;
+      // The following up checkpoints would recommit the instant.
       if (writeClient.getConfig().getFailedWritesCleanPolicy().isLazy()) {
         writeClient.getHeartbeatClient().start(instant);
       }
-      commitInstant(instant);
-    }
-    // stop the heartbeat for old instant
-    if (writeClient.getConfig().getFailedWritesCleanPolicy().isLazy() && !WriteMetadataEvent.BOOTSTRAP_INSTANT.equals(this.instant)) {
-      writeClient.getHeartbeatClient().stop(this.instant);
+      commitInstant(checkpointId, instant, bootstrapBuffer);
     }
-    // starts a new instant
-    startInstant();
     // upgrade downgrade
     this.writeClient.upgradeDowngrade(this.instant, this.metaClient);
   }
 
   private void handleBootstrapEvent(WriteMetadataEvent event) {
-    this.eventBuffer[event.getTaskID()] = event;
+    cleanLegacyEvent(event);
+    WriteMetadataEvent[] eventBuffer = getOrCreateBootstrapBuffer(event);
+    eventBuffer[event.getTaskID()] = event;
     if (Arrays.stream(eventBuffer).allMatch(evt -> evt != null && evt.isBootstrap())) {
-      // start to initialize the instant.
-      final String instant = Arrays.stream(eventBuffer)
-          .filter(evt -> evt.getWriteStatuses().size() > 0)
-          .findFirst().map(WriteMetadataEvent::getInstantTime)
-          .orElse(WriteMetadataEvent.BOOTSTRAP_INSTANT);
-
-      // if currentInstant is pending && bootstrap event instant is empty
-      // reuse currentInstant, reject bootstrap
-      if (this.metaClient.reloadActiveTimeline().filterInflightsAndRequested().containsInstant(this.instant)
-              && instant.equals(WriteMetadataEvent.BOOTSTRAP_INSTANT)
-              && this.tableState.operationType == WriteOperationType.INSERT) {
-        LOG.warn("Reuse current pending Instant {} with {} operationType, "
-                + "ignoring empty bootstrap event.", this.instant, WriteOperationType.INSERT.value());
-        reset();
-
-        // send commit act event to unblock write tasks
-        sendCommitAckEvents(-1L);
-        return;
-      }
+      // start to recommit the instant.
```

Comment:
For partial failover, this condition will not be satisfied, will the event buffer be committed later? e.g., during next ckp complete

Reply:
yes, it will be recommitted later.